### DISCUSSION TITLE
Publishing `clippy_utils` on crates.io

### DISCUSSION THREAD

**Discussion by jubnzv** [rust-lang/rust-clippy#12111] [General]
Hello,

I am reaching out to inquire about the current status of `clippy_utils` and its potential publication on crates.io.

We have a custom crate based on `dylint` for our project, which depends on `clippy_utils`. Currenly, we [use](https://github.com/paritytech/ink/blob/306145c94e78b388511b75ccbe1d77306e8cc49a/linting/extra/Cargo.toml#L21)  `clippy_utils`  that refers to the specific GitHub revision:

```toml
clippy_utils = { git = "https://github.com/rust-lang/rust-clippy", rev = "1d334696587ac22b3a9e651e7ac684ac9e0697b2" }
```

I'm in the process of publishing our crate on crates.io. However, as `cargo publish` requires all dependencies to be available on crates.io, this presents a challenge.

I wanted to reach out and understand the rationale behind the decision [not to publish](https://github.com/rust-lang/rust-clippy/blob/7fbaba856b711f9e945a59a4ba6647a973912564/clippy_utils/Cargo.toml#L5) `clippy_utils` on crates.io. Is there a specific reason for this, and would it be feasible to consider its publication?

Alternatively, could you recommend any solutions that other users have adopted in similar situations?
Thank you for your attention to this matter.

**Comment by flip1995** [comment]
The motivation for not publishing is, that `clippy_utils` is considered just as unstable as `rustc` internals themselves. When creating the `clippy_utils` crate for `dylint` we wanted to be careful not to suggest that `clippy_utils` is stable in any way, which publishing on `crates.io` might do.

Now, `clippy_utils` depends on `clippy_config`, which we also don't really want to publish, because it is a really Clippy internal thing, that is not useful to anyone. This shows, if we were to publish `clippy_utils`, we'd also have to publish all of it deps and future deps.

This adds some maintenance burden, as with every release, we would have to publish every dep of `clippy_utils` and `clippy_utils` itself. We would then need to add a `README.md` to it, in addition to licensing files and some other metadata.

In the end it boils down to us not wanting to provide support for any users of `clippy_utils`, as we don't have the resources to do so.

---
See also https://github.com/rust-lang/rust-clippy/pull/6746#issuecomment-780747522

**Reply by jubnzv** [reply]
Thank you! That seems totally reasonable to me.
